### ISSUE TITLE
Sharing violation on test.stdout.log and test.stderr.log

### ISSUE THREAD

**Issue by gastart** [etas/vs-boost-unit-test-adapter#176]
I am encountering situations where for some yet unknown reason another process is attaching to the test.stdout.log and the test.stderr.log of the files generated by the boost unit test executable.  This is of course causing that the test execution result to become undetermined.  If the user decides to run a batch of tests the whole batch goes in an undetermined state.

The log i have is 

> Executing:   -> [mytest01]
> Exception caught while running test batch \Bin\Debug\Win32\i386\bin\BoostTest01.exe [mytest01] (The process cannot access the file 'C:\Users\gastart\AppData\Local\Temp\BoostTest01.exe.7096.6.test.stdout.log' because it is being used by another process.)

I dunno if we can print out the process with which we are having the access violation violation with so in future we can troubleshoot such problems easier, so as not to have to resolve to other tools like process monitor.

The below settings in the runsettings are a workaround to the problem but they come with the expense of losing functionality.

      <EnableStdOutRedirection>false</EnableStdOutRedirection>
      <EnableStdErrRedirection>false</EnableStdErrRedirection>

**Comment by gastart** [comment]
After deeper investigation of the process monitor logs, a simplified explanation of the problem I encountered is that the problem is that the boost unit test executable enchilada was not 100% ready when the executable returned!!  The test adapter is attempting to read the test.stdout.log about 25ms before all the related processes to the boost unit test executable closed off the test.stdout.log and test.stderr.log, hence causing the access violation.

Inserting a pause in BUTA between the execution phase and the reading off the log file solved the problem.

In essence can you please:

_Insert a configurable sleep time (via the runsettings file) between the execution phase and the start of processing of the log files.  The optional parameter should have a default of zero and is to be defined in milliseconds._

**Comment by guwirth** [comment]
> executable enchilada was not 100% ready when the executable returned!! 
> Insert a configurable sleep time (via the runsettings file) 

Hi @gastart thanks for your investigation. I'm not a big fan of so many configurable properties. Is there no way to ensure that writing the file is finished? Maybe Windows provides such a way, e.g. getting exclusive read access to the file?

**Comment by guwirth** [comment]
**OpenFile** function: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365430%28v=vs.85%29.aspx
with **OF_SHARE_EXCLUSIVE**: Opens a file with exclusive mode, and denies both read/write access to other processes. **If a file has been opened in any other mode for read/write access, even by the current process, the function fails.**

**Comment by netspiri** [comment]
Having debugged the problem with @netspiri, the issue regards **external** processes which are attaching to the test module.

In this particular test scenario, the test module is launching and executing the test successfully. By design, the test adapter redirects the standard output stream to a file. In the meantime, an external process seems to have hooked onto the output stream of the test module.

The problem occurs when the test adapter tries to read the contents of the standard output file while the external process still has a handle open to the stream. As correctly mentioned, [Process Monitor](https://technet.microsoft.com/en-us/sysinternals/processmonitor.aspx?f=255&MSPPError=-2147217396) signals a `SHARING VIOLATION` error, causing the exception to be thrown in the adapter. Apparently, the external process is opening the file utilising this configuration. Analysis using Process Monitor also hint that the external process is not spawned by the test module itself since the parent process is not the test module but this is non-conclusive information.

There were other test runs where the external process relinquished access to the file before the adapter tries to read it. Such cases were successful but, as evident, we are at the mercy of OS scheduling.

Just to be clear, this issue is only apparent in this kind of scenario. In general, this issue can be caused by processes such as anti-virus software, or as in this case, licensing servers. I do not think it is an issue which is perceived by a lot of users (but please, correct me if I am wrong) so starting off with a configuration addition might be a suitable solution for now. If the issue persists, then we can think of a long-er term solution.

**Comment by guwirth** [comment]
@netspiri where in the code do you open the LOG file? Could you provide a link to source code please.
* my idea is still to try OpenFile with OF_SHARE_EXCLUSIVE and if failing try again.
* is there the case where no LOG file exists: means we are running always into the maximum timeout?

**Comment by gastart** [comment]
Hello @guwirth. I gave a quick look at the code and the log file seems to be opened via the methods [GetStandardOutput()](https://github.com/etas/vs-boost-unit-test-adapter/blob/master/BoostTestAdapter/Boost/Results/TestResultCollection.cs#L186) and [GetStandardError()](https://github.com/etas/vs-boost-unit-test-adapter/blob/master/BoostTestAdapter/Boost/Results/TestResultCollection.cs#L206).  The file is being opened in the ctor of [BoostTestResultOutputBase](https://github.com/etas/vs-boost-unit-test-adapter/blob/master/BoostTestAdapter/Boost/Results/BoostTestResultOutputBase.cs#L24)  

I discussed the option of "trying again" with @netspiri before we discovered that the interim sleep would do.  The "trying again" option would require the user defining a number of retries and interim time between retries I guess.

Regarding your second point as far as I know, the user has the option not to generate and process the stdout and stderr via the options

```
<EnableStdOutRedirection>false</EnableStdOutRedirection>
<EnableStdErrRedirection>false</EnableStdErrRedirection>
```

Additionally I think there might be crash situations of the boost unit test executable (for example dll not found) that might lead for the stdout, stderr to not be generated.

**Comment by guwirth** [comment]
@gastart think you are closer to the problem and it's up to you to choose the right solution. My point is only that I'm not a big fan of user options. They are hard to test and maintain (especially combination of settings).

**Comment by gastart** [comment]
I discussed this issue with @netspiri offline.  We agreed that we should implement the sleep period as described in one of my previous comments.  This should be a very minor change in reality and I'll try to work on it sometime in the coming weeks....so please do not block any BUTA releases because of this enhancement.

**Comment by guwirth** [comment]
Closed with #180
